File: code/darwin-systems/src/darwin/sorting/radix-sort/radix_sort.c
/**
 * @file        :   radix_sort.c
 * @author      :   Braiden Gole
 * @brief       :   Implementation of the Radix Sort algorithm.
 * @details     :   This script provides a non-comparative integer sorting 
 * algorithm. It sorts data by processing individual digits 
 * sharing the same significant position and value. This 
 * implementation uses the Least Significant Digit (LSD) approach.
 * @version     :   1.0
 * @date        :   2026-04-17
 */
#include <stdio.h>
#include <stdlib.h>

/**
 * @brief       :   Finds the maximum value in an integer array.
 * @details     :   Iterates through the array to identify the largest element, 
 *                  which determines the number of digits to process.
 * @param       :   array:  The target integer array.
 * @param       :   size:   The number of elements in the array.
 * @return      :   int:    The maximum value found in the array.
 */
static inline int get_maximum(const int* array, size_t size)
{
    int maximum = array[0];

    for (size_t index = 1; index < size; index++)
    {
        if (array[index] > maximum)
        {
            maximum = array[index];
        }
    }

    return maximum;
}

/**
 * @brief       :   Performs a stable counting sort based on a specific digit.
 * @details     :   A helper subroutine that sorts the array based on the digit 
 *                  represented by the current exponent (1, 10, 100, etc.). 
 *                  Stability is maintained to ensure the correctness of the 
 *                  overall Radix Sort.
 * @param       :   array:    The target integer array to be partially sorted.
 * @param       :   size:     The number of elements in the array.
 * @param       :   exponent: The current digit position (10^i) being processed.
 * @return      :   void
 */
static inline void counting_sort(int* array, size_t size, int exponent)
{
    int* output = (int*)malloc(sizeof(int) * size);
    
    if (output == NULL)
    {
        return;
    }

    int count[10] = {0};

    // Store count of occurrences of each digit in count[].
    for (size_t index = 0; index < size; index++)
    {
        count[(array[index] / exponent) % 10]++;
    }

    // Update count[i] to store the actual position of this digit in output[].
    for (int index = 1; index < 10; index++)
    {
        count[index] += count[index - 1];
    }

    // Build the output array using the count positions.
    for (int index = (int)size - 1; index >= 0; index--)
    {
        output[count[(array[index] / exponent) % 10] - 1] = array[index];
        count[(array[index] / exponent) % 10]--;
    }

    // Transfer the sorted elements back into the original array.
    for (size_t index = 0; index < size; index++)
    {
        array[index] = output[index];
    }

    free(output);
}

/**
 * @brief       :   Sorts an array of integers using the Radix Sort algorithm.
 * @details     :   Processes each digit position starting from the least 
 *                  significant digit (LSD) up to the most significant digit. 
 *                  Complexity is O(nk) where k is the number of digits.
 * @param       :   array:  The target integer array to be sorted.
 * @param       :   size:   The number of elements in the array.
 * @return      :   void
 */
void radix_sort(int* array, size_t size)
{
    if (array == NULL || size < 2)
    {
        return;
    }

    // Identify the maximum number to determine the number of digits.
    int maximum = get_maximum(array, size);

    // Apply counting sort to every digit position.
    for (int exponent = 1; maximum / exponent > 0; exponent *= 10)
    {
        counting_sort(array, size, exponent);
    }
}

File: code/golang/radix-sort/radixsort/radix_sort.go
// -----------------------------------------------------------------------------
// Package: radixsort
// File: radixsort.go
//
// Description:
//
//	This file implements a high-performance Radix Sort algorithm for unsigned
//	integers using the Least Significant Digit (LSD) strategy.
//
//	The implementation functions by processing digits from right to left,
//	utilizing a stable Counting Sort as an auxiliary subroutine. This ensures
//	linear time complexity relative to the number of digits while maintaining
//	the relative order of duplicate elements.
//
//	By utilizing fixed-size arrays for digit frequency and minimizing
//	comparisons, this algorithm provides an efficient alternative to
//	comparison-based sorts for large datasets of integers.
//
// -----------------------------------------------------------------------------
package radixsort

import "fmt"

// findMaximum identifies the largest integer within the slice.
// This value is essential for determining the total number of iterations
// required to fully sort the collection.
func findMaximum(items []int) int {
	maximum := items[0]

	for _, number := range items {
		if number > maximum {
			maximum = number
		}
	}

	return maximum
}

// countingSort performs a stable distribution sort on the input slice
// based on the digit represented by the current placeValue.
func countingSort(items []int, placeValue int) {
	size := len(items)
	result := make([]int, size)
	counts := [10]int{}

	// 1. Map digits to their frequency in this digit place.
	for _, number := range items {
		digit := (number / placeValue) % 10
		counts[digit]++
	}

	// 2. Compute prefix sums (cumulative counts).
	// This tells us the ending position for each digit group.
	for index := 1; index < 10; index++ {
		counts[index] += counts[index-1]
	}

	// 3. Build the array.
	// We iterate backwards to maintain "stability" (keep original order).
	for index := size - 1; index >= 0; index-- {
		number := items[index]
		digit := (number / placeValue) % 10

		mappedIndex := counts[digit] - 1
		result[mappedIndex] = number
		counts[digit]--
	}

	// 4. Update the original slice.
	copy(items, result)
}

// RadixSort orchestrates the sorting process by iterating through each
// decimal place of the numbers in the input slice.
func RadixSort(dataset []int) error {
	if len(dataset) == 0 {
		return nil
	}

	// Radix sort digit logic requires non-negative integers.
	// We validate the dataset before processing to prevent incorrect sorting.
	for _, number := range dataset {
		if number < 0 {
			return fmt.Errorf("Negative value %d detected; only non-negative integers supported", number)
		}
	}

	// A slice with only one item is already sorted.
	if len(dataset) < 2 {
		return nil
	}

	maximum := findMaximum(dataset)

	// Place value increments by factors of 10 (1, 10, 100...) to
	// evaluate each digit column until the maximum value is exceeded.
	for placeValue := 1; maximum/placeValue > 0; placeValue *= 10 {
		countingSort(dataset, placeValue)
	}

	return nil
}
